### PR TITLE
Combat replay and Mechanics

### PR DESCRIPTION
**Pull request by LunusLupus**
- Added numerous Mechanics for Mechanics tracking on W6 (especially CA and Qadim)
- Added Combatreplay actors:
  - Sabetha:
    - Kernan Hail of Bullets
    - Knuckles CC
    - Karde Flame Blast
  - Gorseval Slam
  - Slothasor Halitosis (Flame Breath)
  - Cairn Gravity Wave
  - Mursaat Overseer:
    - Jade Scout Shield
    - Jade Explosion
  - Conjured Amalgamate:
    - Shield on Boss and Conjured Shields
    - Purple Protective Area on the ground
  - Twin Largos:
    - CC for both bosses
    - Largos' Domain (Worldeater in CM) for both bosses
    - Aquatic Detainment on Nikare (Floating Bubble)
    - Water puddle on Nikare
    - Shockwave on Kenut
    - Boonsteal on Kenut
  - Qadim:
    - Breakbars on Boss and all Legendary Adds
    - Both Destroyer Shockwaves
    - Destroyer Stomp
    - Hydra Breath
    - Qadim Mace Shockwave
    - Qadim Riposte (CC failure)
    - Wyvern Fire Breath
    - Matriarch Wing Buffet
    - Added Wyvern Tail Swipe
    - Added Destroyer Wave of Force (auto attack)
- Slightly adjusted arena size on the Qadim fight to better show Wyvern phase and last 33% part until dynamic maps are implemented
- Adjusted PieActor for consistency

### CODE REVIEW

File: LuckParser/Models/BossLogic/BossLogic.cs [hunk 1]
```diff
@@ -522,5 +521,11 @@ protected static List<CombatItem> GetFilteredList(ParsedLog log, long skillID, u
             }
             return filtered;
         }
+
+        public static int getRotationFromFacing(Point3D facing)
```
Comment (EliphasNUIT) [inline LuckParser/Models/BossLogic/BossLogic.cs]: I think this method is better off being in Point3D.cs. It looks out of place in boss logic

File: LuckParser/Models/BossLogic/Sabetha.cs [hunk 2]
```diff
@@ -166,8 +169,49 @@ public override void ComputeAdditionalBossData(Boss boss, ParsedLog log)
                     }
                     break;
                 case (ushort)Kernan:
+                    List<CastLog> bulletHail = cls.Where(x => x.SkillId == 31721).ToList();
+                    Boss kernan = Targets.FirstOrDefault(x => x.ID == (ushort)Kernan);
```
Comment (EliphasNUIT) [inline LuckParser/Models/BossLogic/Sabetha.cs]: You should simply be using boss or replay here no? Why look for targets again when everything's already here?

File: LuckParser/Models/BossLogic/Qadim.cs [hunk 3]
```diff
@@ -183,14 +199,238 @@ public override void ComputeAdditionalThrashMobData(Mob mob, ParsedLog log)
...
+                    {
+                        int start = (int)c.Time;
+                        int delay = 2230;
+                        int duration = 2680;
+                        int radius = 2000;
+                        int impactRadius = 40;
+                        Point3D facing = replay.Rotations.LastOrDefault(x => x.Time <= start + 1000);
+                        Point3D bossPosition = replay.Positions.LastOrDefault(x => x.Time <= start + 1000);
+                        int spellCenterDistance = 120; //hitbox radius
+                        Point3D position = new Point3D(bossPosition.X + facing.X * spellCenterDistance, bossPosition.Y + facing.Y * spellCenterDistance, bossPosition.Z, bossPosition.Time);
```
Comment (EliphasNUIT) [inline LuckParser/Models/BossLogic/Qadim.cs]: this should be in the if

Comment (EliphasNUIT) [reply LuckParser/Models/BossLogic/Qadim.cs]: Did you check that facing vector is always normalized?

File: LuckParser/Models/BossLogic/Qadim.cs [hunk 4]
```diff
@@ -183,14 +199,238 @@ public override void ComputeAdditionalThrashMobData(Mob mob, ParsedLog log)
...
+                    {
+                        int start = (int)c.Time;
+                        int delay = 1600;
+                        int duration = 3500;
+                        int maxRadius = 2000;
+                        int impactRadius = 500;
+                        Point3D facing = replay.Rotations.LastOrDefault(x => x.Time <= start + 1000);
+                        Point3D bossPosition = replay.Positions.LastOrDefault(x => x.Time <= start + 1000);
+                        int spellCenterDistance = 270; //hitbox radius
+                        Point3D position = new Point3D(bossPosition.X + facing.X * spellCenterDistance, bossPosition.Y + facing.Y * spellCenterDistance, bossPosition.Z, bossPosition.Time);
```
Comment (EliphasNUIT) [inline LuckParser/Models/BossLogic/Qadim.cs]: this should also be in the if

File: LuckParser/Models/BossLogic/BossLogic.cs (hunk 1 above)
Comment (LunusLupus) [reply LuckParser/Models/BossLogic/BossLogic.cs]: True, moved it.

**Pull request merged**